### ISSUE TITLE
False Positive: BC_UNCONFIRMED_CAST for pattern matching switch

### ISSUE THREAD

**Issue by robinjhector** [spotbugs/spotbugs#2782]
In JDK 21 we can now use pattern matching switches. So instanceof checking can be done in the switch label.

Here's a failing unit test branch that returns `BC_UNCONFIRMED_CAST` (a false positive): https://github.com/spotbugs/spotbugs/compare/master...robinjhector:spotbugs:issue2782

I would be happy to contribute with a fix for this

In-line example

```java
public class Issue2782 {

    public interface MyInterface<T> {
    }

    static class Impl1 implements MyInterface<String> {
        public MyInterface<String> helloFromImpl1() {
            return this;
        }
    }

    static class Impl2 implements MyInterface<Boolean> {
        public MyInterface<Boolean> helloFromImpl2() {
            return this;
        }
    }

    public static <T> void method(MyInterface<T> instance) {
        Object bla = switch (instance) {
            case Impl1 impl1 -> impl1.helloFromImpl1();
            case Impl2 impl2 -> impl2.helloFromImpl2();
            default -> throw new IllegalArgumentException("Blah");
        };

        System.out.println(bla);
    }
}
```

**Comment by robinjhector** [comment]
I can't get the development environment to work locally with gradle for some reason, so I can't pinpoint where the false-positive occurs. Hopefully someone else can take a look at this!

**Comment by robinjhector** [comment]
This also applies to `BC_UNCONFIRMED_CAST_OF_RETURN_VALUE`, even though the return value is never cast to anything

**Comment by gtoison** [comment]
> I can't get the development environment to work locally with gradle for some reason

What's the error you're getting? We're looking for ways to help external contributors onboarding into the project

**Comment by robinjhector** [comment]
> > I can't get the development environment to work locally with gradle for some reason
> 
> What's the error you're getting? We're looking for ways to help external contributors onboarding into the project

It compiled fine iirc, but I couldn't get the debugger to attach correctly in IntelliJ. No breakpoints were respected. I don't know how to trouble shoot it since I'm more of a Maven guy myself, and has rarely worked with gradle

**Comment by maxl2287** [comment]
Hi,
is it already resolved?
Because with 4.8.5.0 (maven-spotbugs) I still got this spotbug for:

```java
boolean isValid = switch (geographicArea) {
      case Polygon polygon -> ...
      case PointUncertaintyCircle circle -> ...
      case Point point -> ...
      default -> {
        yield true;
      }
    };
```

`[ERROR] Medium: Unchecked/unconfirmed cast`

**Comment by robinjhector** [comment]
> Hi, is it already resolved? Because with 4.8.5.0 (maven-spotbugs) I still got this spotbug for:

The fix will be included in 4.8.6 by the looks of it 👍

**Comment by gtoison** [comment]
So this should be fixed now that 4.8.6 is released
